Update FileSystemProvider config info for V8 (#2023)
umbraco/UmbracoDocs: Extending/FileSystemProviders/index.md

---
versionFrom: 8.0.0
---

# Custom file systems (IFileSystem)

By default, Umbraco uses an instance of `PhysicalFileSystem` to handle the storage location of the media archive (~/media).

This can be configured during composition:

```csharp
using Umbraco.Core.Composing;
using Umbraco.Core;
using Umbraco.Core.IO;

namespace Umbraco8Examples.Composition
{
    public class SetMediaFileSystemComposer : IUserComposer
    {
        public void Compose(Umbraco.Core.Composing.Composition composition)
        {
            composition.SetMediaFileSystem(() => new PhysicalFileSystem("~/custommediafolder"));
        }
    }
}
```

> The PhysicalFileSystem accepts the application path to the 'virtual Root' to location the Media folder the ~ is important!

Author: @sofietoft yes, it is a bit funky.. have updated in latest push

Alternatively you can combine:
a **RootPath** - the rooted, filesystem path, using directory separator chars, NOT ending with a separator // eg "c:" or "c:\path\to\site" or "\\server\path"
and **RootUrl** - the relative url, using url separator chars, NOT ending with a separator // eg "" or "/Views" or "/Media" or "/<vpath>/Media" in case of a virtual path

> 

Reviewer: To the above:
As I understand it, you can **combine** `RootPath` and `RootUrl`, right?
Would it make sense to add an example maybe? 🤔

Author: @sofietoft added in latest push...

:::

`PhysicalFileSystem` implements the `IFileSystem` interface, and it is possible to replace it with a custom class - eg. if you want your media files stored on Azure or something similar.

If you configure Umbraco to use a custom file system provider for media, you most likely won't need to access the implementation directly. Umbraco uses a wrapper class called `MediaFileSystem`. You can get a reference to this wrapper class with the following code:

```csharp
IMediaFileSystem media = Current.MediaFileSystem;
```
or via dependency injection in the constructor for your custom class or controller:
```csharp
 public class ImagesController : UmbracoAuthorizedApiController
    {
        private readonly IMediaFileSystem _mediaFileSystem;

        public ImagesController(IMediaFileSystem mediaFileSystem)
        {
            _mediaFileSystem = mediaFileSystem;
        }
```

This will be enough in most cases and is the way Umbraco will access the file system provider. The wrapper class implements the same interface `IFileSystem` as any custom providers should do, so you will be able to call the same methods.

> ## Setting the PhysicalFileSystem to be outside of the application root

Reviewer: Is this heading a heading that includes `## MediaPath Scheme` and the rest of the `##` headings in this article? 🤔

Author: ahh sorry yes should be removed as i incorporated that bit further up... updated

## MediaPath Scheme

The MediaPath Scheme defines the current set of rules that decide the format of the Media Path when it is saved into the media archive wherever it is located.

> By default the MediaPath scheme used by Umbraco is the 'UniqueMediaPathScheme' this generates a unique 'folder' to place the uploaded image in eg

Reviewer: ```suggestion
By default the MediaPath scheme used by Umbraco is the `UniqueMediaPathScheme` this generates a unique 'folder' to place the uploaded image in eg.
```

> /media/dozdrg2f/mylovelyimage.jpg

Reviewer: ```suggestion
`/media/dozdrg2f/mylovelyimage.jpg`
```

> '/media' is defined by the PhysicalFileSystem and 'dozdrg2f' is generated by the UniqueMediaPathScheme.

Reviewer: ```suggestion
`/media` is defined by the PhysicalFileSystem and 'dozdrg2f' is generated by the `UniqueMediaPathScheme`.
```

:::note
In Umbraco 7 the integer ids were used in the path, and this approach is still possible using the 'OriginalMediaPathScheme'
:::

> are stored in the format ~/media/

Reviewer: Is this sentence supposed to be inside the note? 🤔

Author: just shouldn't be there :-)

> You can set the MediaPathScheme during composition, for example if you wanted to revert back to the V7 methodology in a migrated site:

Reviewer: ```suggestion
You can set the `MediaPathScheme` during composition, for example if you wanted to revert back to the V7 methodology in a migrated site:
```

or again via dependency injection

```csharp
  public class FileSystemLocations 
    {
        private readonly IFileSystems _fileSystems;
        public FileSystemLocations(IFileSystems fileSystems)
        {
            _fileSystems = fileSystems;
            var macroPartialsFileSystem = _fileSystems.MacroPartialsFileSystem;
        }        
```

Both `IFileSystem` and `IMediaFileSystem` are located in the `Umbraco.Core.IO` namespace.

## Custom providers

There is an Azure Blob Storage provider:

* [Azure Blob Storage](Azure-Blob-Storage/)

### Creating your own FileSystem provider
Create your filesystem class:
```csharp
            public class MyFileSystem : FileSystemWrapper
            {
                public MyFileSystem(IFileSystem innerFileSystem)
                    : base(innerFileSystem)
                { }
            }
```
The constructor can have more parameters, that will be resolved by the dependency injection container.

Register your new filesystem implementation, in a component:
```csharp
  composition.RegisterFileSystem<MyFileSystem>();
```
Register the underlying filesystem:
```csharp
  composition.RegisterUniqueFor<IFileSystem, MyFileSystem>(...);
```

> You can inject MyFileSystem wherever it's needed.

Reviewer: ```suggestion
You can inject `MyFileSystem` wherever it's needed.
```

Furthermore it's possible to declare a filesystem interface:
```csharp
  public interface IMyFileSystem : IFileSystem
  { }
```
Make the class implement the interface, then
register your filesystem, in a component:
```csharp
  composition.RegisterFileSystem<IMyFileSystem, MyFileSystem>();
  composition.RegisterUniqueFor<IFileSystem, IMyFileSystem>(...);
```
You can inject IMyFileSystem wherever it's needed.

> ### What is Shadowing?

Reviewer: Is this something you'll run into when working with this? 
Sounds like it's an Umbraco Cloud thing? 🤔 Not entirely sure I understand the context here.. but I might be missing something very obvious 😅

Author: I have no idea, I found the explanation in the source code following the adivce on IMyFileSystem... I have no idea if it helps to know it... I reckon we take it out?

Author: I have removed it

Reviewer: Yes, I agree. It did confuse me a bit, so it might confuse others as well (maybe!) 😅
